BinStat

kreisl opened the pull request:
New Statistics objects which can be created from Stats object. This should fix shortcomings of the old approach.
Drawing not yet implemented, but should be straight forward.

kreisl commented:
Can you have a look if the class give us what we need? We could then also merge it with the other implementation of Stats Statistic and Resamples. Or keep them separate one class for gathering statistics and one for calculation of observables?

I used the arithmetics shown https://en.wikipedia.org/wiki/Algebra_of_random_variables and here https://link.springer.com/chapter/10.1007/978-1-4939-6572-4_4#Tab1 . We don't need to keep track of the moments to merge bins, I implemented the merging using the weighted average. This should also correctly merge the variance.
The formula is available here https://en.wikipedia.org/wiki/Pooled_variance#Sample-based_statistics .

mam-mih-val commented:
@kreisl, I checked this on real data set. Rebin and projection seems to be working. But I've found, that 
1. After rebinnig and projection, errors in bins are zero. Should it be like this? 
2. If I do the same operation for instance calculation 3-sub resolution with stats and stat bins, errors from bootstrapping differs from each other

kreisl commented:
1. Yes, I fixed it locally already. Will push changes after I resolve 2.
2. I think I accidentally calculated the variance of the sample not the variance of the sample mean, which is what we want. I need to fix the formulas. something like multiplying by (sum w)^2/ sum w^2

kreisl commented:
Latest push should have fixed the issues seen previously, if I understood them correctly.

mam-mih-val commented:
1D-container of stat bins works correctly, checked rebinnig and compared with initital container of stats. But seems to be one more bug appeared: rebinnig ang projection of multidimentional data container of stat bins refused to work. If I do rebinnig, it throws me that I have overlapping bins, even if I try to merge all them in one sigle bin through axis. And if I try to do projection, I get "nan" in means of bins, but error estimated normally.

kreisl commented:
I added a 2-dimensional test case. I cannot reproduce the issue you are seeing. Are you just using this branch on top of master?

eugene274 commented on src/base/DataContainer.test.cpp:
```diff
@@ -67,11 +72,17 @@ TEST_F(DataContainerUnitTest, StatBinMerging) {
+  merged_statbin.SetErrors(Qn::StatBin::ErrorType::PROPAGATION);
+  auto prop_error = merged_statbin[0].StdDevOfMeanFromPropagation();
+  merged_statbin.SetErrors(Qn::StatBin::ErrorType::BOOTSTRAP);
+  auto bs_error = merged_statbin[0].StdDevOfMeanFromPropagation();
```
It seems to be a copy-paste error

eugene274 commented on src/base/DataContainer.test.cpp:
```diff
@@ -81,6 +92,12 @@ TEST_F(DataContainerUnitTest, StatBinAddition) {
+  sum.SetErrors(Qn::StatBin::ErrorType::PROPAGATION);
+  auto prop_error = sum[0].StdDevOfMeanFromPropagation();
+  sum.SetErrors(Qn::StatBin::ErrorType::BOOTSTRAP);
+  auto bs_error = sum[0].StdDevOfMeanFromPropagation();
```
also copy-paste error?

kreisl replied:
yes i fixed them all

kreisl replied:
thanks 👍

mam-mih-val commented:
I'm sorry, that was my bad with overlapping bins, I set rebin range incorrectly. But still, I get "nan" mean and error of stat bins in case of projection multidimensional container of stat bins on single axis. I assume it might be because of neighboring empty bins in initial container. I mean, this bins are observable, which means we should take into account their weights, but somehow, attempt to merge two empty bins products nan.

eugene274 commented on include/base/QnTools/StatBin.hpp:
```diff
@@ -0,0 +1,220 @@
+
+#include <cmath>
+#include <vector>
+
```
Missing headers: Rtypes.h, Stats.hpp

kreisl replied:
added

eugene274 commented on include/base/QnTools/StatBin.hpp:
```diff
@@ -0,0 +1,220 @@
+  StatBin merged;
+  // mean and variance
+  merged.weight_ = lhs.Weight() + rhs.Weight();
+  merged.mean_ = (lhs.Mean() * lhs.Weight() + rhs.Mean() * rhs.Weight()) / merged.Weight();
```
if `lhs.Mean()` or `rhs.Mean()` is _nan_, `merged.mean_` gets _nan_.

Test case:
* merge two empty StatBin-s
* merge result with filled StatBin

kreisl replied:
implemented a check based on empty weights

eugene274 commented:
@kreisl important offtop. When we discussed our Git workflow, you proposed keeping our development branches in private forks. I was rooting for having development branches in HeavyIonAnalysis/QnTools. You've submitted pull request. I wanted to add unit test reproducing Mikhail's issue. I cannot do that in you PR, because that is inside your private fork.

kreisl commented:
@eugene274  as a maintainer you can push to other peoples pull requests. I think this is a good way of how to handle this procedure. For example: `git push git@github.com:user/repo local_branch_name:remote_branch_name`
should work

eugene274 commented:
> @eugene274 as a maintainer you can push to other peoples pull requests. I think this is a good way of how to handle this procedure. For example: `git push git@github.com:user/repo local_branch_name:remote_branch_name`
> should work

And it really works! Question closed. Thanks!

kreisl commented:
Made another push cleaning up the classes a bit and fixing the bugs you have found. I notice the bs error seem to underestimate the statistical error. In some cases quite a bit. That's why the tests fail for some of the arithmetic operations. I am not sure if there is still a bug in the implementation or what deviation should be expected

mam-mih-val commented:
@kreisl, I checked this out on real data sample again. It became better, means looks ok, but still in several bins there are "nan" errors. Checked error type. It's bootstrap. I think it's still the problem with merging empty bins.

mam-mih-val commented:
@kreisl, propose: in StatBin::MergeBins() add special case for empty lhs and rhs in the same time, beacause situation improved, but somehow, some of bin errors still are nan.

kreisl commented on src/base/StatBin.cpp:
```diff
@@ -48,9 +48,12 @@ StatBin Merge(const StatBin &lhs, const StatBin &rhs) {
+  } else if (rhs.sum_weight_ == 0.) {
+    merged.sample_means_ = lhs.sample_means_;
+    merged.sample_weights_ = lhs.sample_weights_;
   } else {
```
the only time when the division by sum of weights happens is when one of the bins which are being merged is not empty. so I am not sure why nan is still popping up?

kreisl replied:
> @kreisl, propose: in StatBin::MergeBins() add special case for empty lhs and rhs in the same time, beacause situation improved, but somehow, some of bin errors still are nan.

Can you send me the container before merging? I can have a look and try to debug.

mam-mih-val replied:
Okay, I'm sending you file with containers via email.

kreisl commented:
There are Problems when merging bins with the variance from propagation. We need to fix the case where the weights are unequal 1. I am not sure which is the right formula. The variance from bootstrapping should not be affected by this problem.

kreisl commented:
I made some new classes similar to what we discussed.
All of the accumulation of the statistics is happening using the StatCollect class.
It has two important members:
1. Statistics, which takes care of the summary statistics.
2. Bootstrap, which takes care of bootstrapping the standard error of the mean.
When we want to calculate an observable, or load the DataContainers containing the StatCollect from a files, we just Create a new one using `DataContainer<StatCalculate>(DataContainer<StatCollect> &)`. With this new container all of the operation to construct the variables can be performed.

kreisl committed:
BinStat

added new Statistics object which should fix the shortcomings of the old one. Much simpler can be constructed form Stats object.
clean up unittests


Added reference and observable type to take into account the weights


Added drawing and fixed error


Added bootstrapping to propagation error comparison to tests


added 2d unit test for statbin


Port tests from #26

fixed bugs and split class

added contributors to docs
fixed empty bins


removing bias correction in merging

usually we use big samples so the bias correction should be small. The formula that was used before was causing nan with certain weights, so it was not correct...
move tests to StatBin.test.cpp

fix nan issues of empty statistics
fixed variance when scaling by constant


Added Statistic unit tests

fixed warning in QVector
added bootstrapping functionality to Statistic

as a move to remove the Stats and ReSample classes. I think they can be removed by a more powerful Statistic class with the bootstrapping means and weights as first class members. We can remove two classes for slightly more complex Statistic class. Workflow will be fill into Statistic -> transform to StatBin -> calculate observable
removed Stats from CorrelationAction


replaced Stats with Stat classes

- Stat contains the flags and is the base of: 
1. StatCollect for collecting the statistics.
2. StatCalculate for calculating observables.
- Statistics only contains the summary statistics
- Bootstrap contains the bootstrap samples

fixed drawing

standardised naming of statistics classes
Clean up drawing functions


Add bootstrap streamer
